### PR TITLE
[FIXED JENKINS-40912] Add method for getting stash file list

### PR DESCRIPTION
**Pull request by abayer**
[JENKINS-40912](https://issues.jenkins-ci.org/browse/JENKINS-40912)

Replaces busted #56 

cc @reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/workflow/flow/StashManager.java [hunk 1]
```diff
@@ -165,6 +167,35 @@ public static void copyAll(@Nonnull Run<?,?> from, @Nonnull Run<?,?> to) throws
...
+     * @throws IOException If unable to interact with the stash files.
+     */
+    @Nonnull
+    public static List<String> stashFileList(@Nonnull Run<?,?> build, @Nonnull String stashName) throws IOException {
+        List<String> contents = new ArrayList<>();
+        File[] kids = storage(build).listFiles();
+        if (kids != null) {
+            for (File kid : kids) {
+                String n = kid.getName();
+                if (n.endsWith(SUFFIX) && stashName.equals(n.substring(0, n.length() - SUFFIX.length()))) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/flow/StashManager.java]: Would be easier to look up the expected child name directly: `storage(build, stashName)`

File: src/main/java/org/jenkinsci/plugins/workflow/flow/StashManager.java [hunk 2]
```diff
@@ -165,6 +167,30 @@ public static void copyAll(@Nonnull Run<?,?> from, @Nonnull Run<?,?> to) throws
...
+    /**
+     * Get the filenames included in a particular stash.
+     * @param build The build to look in.
+     * @param stashName The stash name to look up.
+     * @return A possibly empty list of filenames included in the given stash name in the given build. Unix separators will be used.
+     * @throws IOException If unable to interact with the stash files.
+     */
+    @Nonnull
+    public static List<String> stashFileList(@Nonnull Run<?,?> build, @Nonnull String stashName) throws IOException {
+        List<String> contents = new ArrayList<>();
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/flow/StashManager.java]: Why not `Set<String>`?

File: src/test/java/org/jenkinsci/plugins/workflow/flow/StashManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+                List<String> secondList = StashManager.stashFileList(b, "second");
+                assertFalse(secondList.isEmpty());
+                assertEquals(1, secondList.size());
+                assertTrue(secondList.contains("elsewhere/nested"));
+            }
+        });
+    }
+
+    private static final class HasWorkspacePredicate implements Predicate<FlowNode> {
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/flow/StashManagerTest.java]: There is a utility for this in `workflow-api`.

File: src/main/java/org/jenkinsci/plugins/workflow/flow/StashManager.java (hunk 2 above)
Comment (svanoort) [reply src/main/java/org/jenkinsci/plugins/workflow/flow/StashManager.java]: `Set` is probably more technically correct, but I think `List` will be easier to work with due to positional access.

File: src/main/java/org/jenkinsci/plugins/workflow/flow/StashManager.java [hunk 4]
```diff
@@ -165,6 +167,30 @@ public static void copyAll(@Nonnull Run<?,?> from, @Nonnull Run<?,?> to) throws
         FileUtils.copyDirectory(fromStorage, storage(to));
     }
 
+    /**
+     * Get the filenames included in a particular stash.
```
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/workflow/flow/StashManager.java]: Should include a hard warning that this extracts the whole stash -- this is not a free or cheap operation.

File: src/test/java/org/jenkinsci/plugins/workflow/flow/StashManagerTest.java (hunk 3 above)
Comment (svanoort) [reply src/test/java/org/jenkinsci/plugins/workflow/flow/StashManagerTest.java]: Yup, `FlowScanningUtils.hasActionPredicate(Class<? extends Action)` IIRC  🐜

**Pull request closed without merging**